### PR TITLE
feat(cloudfront): write Lambda@Edge with typescript

### PR DESCRIPTION
**Pull request by mmuller88**
fixes: #12671

As you can see I duplicated a lot from aws-lambda-nodejs/lib/function.ts . It would be great to get rid of that. Let me know how to proceed.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by mmuller88** [comment]
Will start writing the tests if that PR hits interested and after we are happy the the production code part.

**Comment by mmuller88** [comment]
@comcalvi  can I have some feedback for how to get rid of those duplicates?

**Comment by mmuller88** [comment]
@rix0rrr @comcalvi what is going on here?

File: packages/@aws-cdk/aws-cloudfront/lib/experimental/nodejs-edge-function.ts [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+      break;
+    }
+  }
+
+  if (!definingIndex || !sites[definingIndex]) {
+    throw new Error('Cannot find defining file.');
+  }
+
+  return sites[definingIndex].getFileName();
+}
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-cloudfront/lib/experimental/nodejs-edge-function.ts:208]: Could we get a blank line at EOF?

**Review by comcalvi — CHANGES_REQUESTED** [review]
Thanks for the submission. This will need some changes before we can merge it in.

**Comment by mmuller88** [comment]
@comcalvi thx a lot for looking into that. But please read and try to understand the description as that will help to answer most of your questions!

**Comment by comcalvi** [comment]
@mmuller88 before we get into the details of the code duplication, please update the PR so that the code build job succeeds.

**Comment by mmuller88** [comment]
@comcalvi I didn't feel well supported by you. You didn't understand or looked into much of what I was asking for or didn't understand it and clarified that. You are perhaps quite Junior so then I could understand that. A bit disappointed with you and AWS.

File: packages/@aws-cdk/aws-cloudfront/lib/experimental/nodejs-edge-function.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+/**
+ * Searches for an entry file. Preference order is the following:
+ * 1. Given entry file
+ * 2. A .ts file named as the defining file with id as suffix (defining-file.id.ts)
+ * 3. A .js file name as the defining file with id as suffix (defining-file.id.js)
+ * 4. A .mjs file name as the defining file with id as suffix (defining-file.id.mjs)
+ */
+function findEntry(id: string, entry?: string): string {
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply packages/@aws-cdk/aws-cloudfront/lib/experimental/nodejs-edge-function.ts:157]: That's what NodejsFunction does, but the `findEntry` helper is not exposed for reuse. It makes sense to keep the same behavior for both NodejsFunction and NodejsEdgeFunction.

File: packages/@aws-cdk/aws-cloudfront/lib/experimental/nodejs-edge-function.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+export interface NodejsEdgeFunctionProps extends lambda.FunctionOptions { //extends Omit<lambda.FunctionOptions, 'environment'> {
+  /**
+ * Path to the entry file (JavaScript or TypeScript).
+ *
+ * @default - Derived from the name of the defining file and the construct's id.
+ * If the `NodejsFunction` is defined in `stack.ts` with `my-handler` as id
+ * (`new NodejsFunction(this, 'my-handler')`), the construct will look at `stack.my-handler.ts`
+ * and `stack.my-handler.js`.
+ */
+  readonly entry?: string;
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply packages/@aws-cdk/aws-cloudfront/lib/experimental/nodejs-edge-function.ts:24]: The NodejsFunction does the same as well: it looks for a specific entry file based on the file where the NodejsFunction is created.

**Comment by skyrpex** [comment]
Similar issue here https://github.com/aws/aws-cdk/issues/14215

**Comment by mmuller88** [comment]
@comcalvi sorry for my stupid words. I think I was a bit frustrated and acted quite childish. I should have acted on your comments better and with more patience. I think I was frustrated that it took you so long for looking over my issue. But of course, the CDK team has lots to do. Again, I am very sorry.

**Comment by TheOrangePuff** [comment]
@comcalvi @skyrpex I'm happy to take this PR on. This is a huge problem for us and is holding us back from upgrading to CDK 2.

**Comment by comcalvi** [comment]
sorry all! I hadn't seen these since the PR was autoclosed. 

@mmuller88 @TheOrangePuff do you want me to re-open this PR?

**Comment by hugomallet** [comment]
@comcalvi Yes please I think there's still no solution to this

**Pull request closed without merging**